### PR TITLE
[DOC] More documentation for HIBF and config.

### PR DESCRIPTION
**Pull request by smehringer**

### CODE REVIEW

File: include/hibf/hierarchical_interleaved_bloom_filter.hpp [hunk 1]
```diff
@@ -28,48 +28,58 @@
...
- * bins are unevenly distributed.
+ * to establish a hierarchical structure. It is especially suited if you want to index are many samples/sets/user bins
+ * or if their sizes are unevenly distributed.
  *
- * # Terminology
+ * Publication reference: https://doi.org/10.1186/s13059-023-02971-4
  *
- * ## Technical Bin
- * A Technical Bin represents an actual bin in the binning directory. In the IBF, it stores its kmers in a single Bloom
- * Filter (which is interleaved with all the other BFs).
```
Comment (smehringer) [inline include/hibf/hierarchical_interleaved_bloom_filter.hpp:48]: I moved the technical details of the data structure to the bottom.

I think most people are interested in `how` it works first. Later they can read up `why`

File: include/hibf/config.hpp [hunk 2]
```diff
@@ -26,15 +26,33 @@ namespace seqan::hibf
...
+ * | Layout  | seqan::hibf::config::sketch_bits                 | 12      |                        |
+ * | Layout  | seqan::hibf::config::tmax                        | 0       | 0 indicates unset      |
+ * | Layout  | seqan::hibf::config::max_rearrangement_ratio     |         |                        |
+ * | Layout  | seqan::hibf::config::alpha                       | 1.2     |                        |
+ * | Layout  | seqan::hibf::config::disable_estimate_union      | false   |                        |
+ * | Layout  | seqan::hibf::config::disable_rearrangement       | false   |                        |
+ *
+ * As a copy and paste source, here are all config options with their defaults:
+ *
+ * \include test/snippet/hibf_construction.cpp
```
Comment (eseiler) [inline include/hibf/config.hpp:55]:
```suggestion
 * \include test/snippet/hibf/hibf_construction.cpp
```

File: include/hibf/config.hpp [hunk 3]
```diff
@@ -26,15 +26,33 @@ namespace seqan::hibf
...
+ * | Type    |  Option Name                                     | Default | Note                   |
+ * |:--------|:-------------------------------------------------|:-------:|:-----------------------|
+ * | General | seqan::hibf::config::input_fn                    | -       | [REQUIRED]             |
+ * | General | seqan::hibf::config::number_of_user_bins         | -       | [REQUIRED]             |
+ * | General | seqan::hibf::config::number_of_hash_functions    | 2       |                        |
+ * | General | seqan::hibf::config::maximum_false_positive_rate | 0.05    | [RECOMMENDED_TO_ADAPT] |
+ * | General | seqan::hibf::config::threads                     | 1       | [RECOMMENDED_TO_ADAPT] |
+ * | Layout  | seqan::hibf::config::sketch_bits                 | 12      |                        |
+ * | Layout  | seqan::hibf::config::tmax                        | 0       | 0 indicates unset      |
+ * | Layout  | seqan::hibf::config::max_rearrangement_ratio     |         |                        |
```
Comment (eseiler) [inline include/hibf/config.hpp:48]:
```suggestion
 * | Layout  | seqan::hibf::config::max_rearrangement_ratio     | 0.5     |                        |
```

File: include/hibf/hierarchical_interleaved_bloom_filter.hpp [hunk 4]
```diff
@@ -28,48 +28,58 @@
...
- * seqan::hibf::hierarchical_interleaved_bloom_filter::membership_agent() and use the returned
+ * Please see the documentation of seqan::hibf::config for details on how to configure the HIBF construction.
+ *
+ * ## Membership Queries with the HIBF
+ *
+ * To allow efficient, thread-safe membership queries, you need to use the
  * seqan::hibf::hierarchical_interleaved_bloom_filter::membership_agent.
- * In contrast to the [seqan::hibf::interleaved_bloom_filter][1], the result will consist of indices of user bins.
+ *
+ * \include test/snippets/hibf/hierarchical_interleaved_bloom_filter.cpp
```
Comment (eseiler) [inline include/hibf/hierarchical_interleaved_bloom_filter.hpp:66]:
```suggestion
 * \include test/snippet/hibf/hierarchical_interleaved_bloom_filter.cpp
```

File: include/hibf/hierarchical_interleaved_bloom_filter.hpp [hunk 5]
```diff
@@ -28,48 +28,58 @@
...
  *
- * The individual IBFs may have a different number of technical bins and differ in their sizes, allowing an efficient
- * distribution of the user bins.
+ * The HIBF is constructed by passing a seqan::hibf::config. There are two options required to be set:
+ * (1) seqan::hibf::config::input_fn and (2) seqan::hibf::config::number_of_user bins. For all other options
+ * we have set sensible defaults.
+ *
+ * Here are all options with their defaults:
+ *
+ * \input test/snippet/hibf/hibf_construction.cpp
```
Comment (eseiler) [inline include/hibf/hierarchical_interleaved_bloom_filter.hpp:57]:
```suggestion
 * \include test/snippet/hibf/hibf_construction.cpp
```

File: include/hibf/hierarchical_interleaved_bloom_filter.hpp [hunk 6]
```diff
@@ -28,48 +28,58 @@
...
  *
- * ## Technical Bin
- * A Technical Bin represents an actual bin in the binning directory. In the IBF, it stores its kmers in a single Bloom
- * Filter (which is interleaved with all the other BFs).
+ * ### Example
  *
- * ## User Bin
- * The user may impose a structure on his sequence data in the form of logical groups (e.g. species). When querying the
- * IBF, the user is interested in an answer that differentiates between these groups.
+ * \include test/snippets/hibf/hierarchical_interleaved_bloom_filter.cpp
```
Comment (eseiler) [inline include/hibf/hierarchical_interleaved_bloom_filter.hpp:42]:
```suggestion
 * \include test/snippet/hibf/hierarchical_interleaved_bloom_filter.cpp
```

File: test/snippet/hibf/hibf_construction.cpp [hunk 7]
```diff
@@ -0,0 +1,30 @@
...
+
+    seqan::hibf::config config{.input_fn = my_input,     // required
+                               .number_of_user_bins = 2, // required
+                               .number_of_hash_functions = 2,
+                               .maximum_false_positive_rate = 0.05, // recommended to adapt
+                               .threads = 1,                        // recommended to adapt
+                               .sketch_bits = 12,
+                               .tmax = 0, // triggers default copmutation
+                               .max_rearrangement_ratio = 0.5,
+                               .alpha = 1.2,
```
Comment (eseiler) [inline test/snippet/hibf/hibf_construction.cpp:24]:
```suggestion
                               .alpha = 1.2,
                               .max_rearrangement_ratio = 0.5,
```

File: test/snippet/hibf/hierarchical_interleaved_bloom_filter.cpp [hunk 8]
```diff
@@ -0,0 +1,28 @@
...
+    // construct the HIBF
+    seqan::hibf::hierarchical_interleaved_bloom_filter hibf{config};
+
+    // query the HIBF
+    std::vector<size_t> query{1u, 2u, 3u};
+    std::vector<size_t> query2{8u, 9u, 10u};
+
+    auto agent = hibf.membership_agent();            // you need an agent for efficient queries
+    auto & result = agent.membership_for(query, 2u); // result = [0, 1]; both user bins have hashes 1,2,3
+    auto & result = agent.membership_for(query, 2u); // result = [0]; only user bin 0 has hashes 8,9,10
```
Comment (eseiler) [inline test/snippet/hibf/hierarchical_interleaved_bloom_filter.cpp:27]:
```suggestion
    result = agent.membership_for(query, 2u); // result = [0]; only user bin 0 has hashes 8,9,10
```

File: test/snippet/hibf/hierarchical_interleaved_bloom_filter.cpp [hunk 9]
```diff
@@ -0,0 +1,28 @@
...
+    // construct the HIBF
+    seqan::hibf::hierarchical_interleaved_bloom_filter hibf{config};
+
+    // query the HIBF
+    std::vector<size_t> query{1u, 2u, 3u};
+    std::vector<size_t> query2{8u, 9u, 10u};
+
+    auto agent = hibf.membership_agent();            // you need an agent for efficient queries
+    auto & result = agent.membership_for(query, 2u); // result = [0, 1]; both user bins have hashes 1,2,3
+    result = agent.membership_for(query, 2u);        // result = [0]; only user bin 0 has hashes 8,9,10
```
Comment (eseiler) [inline test/snippet/hibf/hierarchical_interleaved_bloom_filter.cpp:27]:
```suggestion
    auto & result2 = agent.membership_for(query, 2u);        // result2 = [0]; only user bin 0 has hashes 8,9,10
```

File: test/snippet/hibf/hierarchical_interleaved_bloom_filter.cpp [hunk 10]
```diff
@@ -0,0 +1,46 @@
...
+    std::vector<size_t> query{1u, 2u, 3u};
+    std::vector<size_t> query2{8u, 9u, 10u};
+
+    auto agent = hibf.membership_agent();             // you need an agent for efficient queries
+    auto & result = agent.membership_for(query, 2u);  // both user bins have hashes 1,2,3
+    print(result);                                    // [1,0]
+    agent.sort_results();                             // Results can also be sorted
+    print(result);                                    // [0,1]
+    auto & result2 = agent.membership_for(query, 2u); // only user bin 0 has hashes 8,9,10
+    print(result2);                                   // [0]
```
Comment (eseiler) [inline test/snippet/hibf/hierarchical_interleaved_bloom_filter.cpp:45]:
```suggestion
    auto agent = hibf.membership_agent();              // you need an agent for efficient queries
    auto & result = agent.membership_for(query, 2u);   // both user bins have hashes 1,2,3
    print(result);                                     // [1,0]
    agent.sort_results();                              // Results can also be sorted
    print(result);                                     // [0,1]
    auto & result2 = agent.membership_for(query2, 2u); // only user bin 0 has hashes 8,9,10
    print(result2);                                    // [0]
```

File: test/snippet/hibf/hierarchical_interleaved_bloom_filter.out [hunk 11]
```diff
@@ -0,0 +1,3 @@
+[1,0]
+[0,1]
+[1,0]
```
Comment (eseiler) [inline test/snippet/hibf/hierarchical_interleaved_bloom_filter.out:3]:
```suggestion
[0]
```

**Review by eseiler — APPROVED** [review]

**Pull request merged**